### PR TITLE
Update Hive Metastore tables for Spark

### PR DESCRIPTION
**Pull request by rdblue**
These changes are needed to add Hive table support to the Spark integration.

* Spark uses Hive 1.2.1, so change the version to 1.2.1
* Refactor tests to share a metastore instance
* Use a client connection pool to avoid too many connections to the metastore to fix failing tests when using a shared metastore instance

### CODE REVIEW

**Comment by rdblue** [comment]
@Parth-Brahmbhatt, can you review these Hive changes?

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 1]
```diff
@@ -231,8 +242,11 @@ private long acquireLock() throws UnknownHostException, TException {
   private void unlock(Optional<Long> lockId) {
     if (lockId.isPresent()) {
       try {
-        metaStoreClient.unlock(new UnlockRequest(lockId.get()));
-      } catch (TException e) {
+        metaClients.run(client -> {
+          client.unlock(lockId.get());
+          return null;
+        });
+      } catch (Exception e) {
```
Comment (rdblue) [inline hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: TODO: This should be `InterruptedException`

**Comment by rdblue** [comment]
@aokolnychyi, you're probably interested in these changes as well. This adds a metastore client pool like you suggested. It also updates your concurrent Hive test to use a barrier instead of trying to sleep the right amount. That way if a commit takes longer, synchronization still happens. Please have a look.

**Comment by aokolnychyi** [comment]
@rdblue thanks for pinging me. We are definitely interested in this, I'll take a look today/tomorrow.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTables.java [hunk 2]
```diff
@@ -70,12 +71,17 @@ public Table load(String tableIdentifier) {
...
+
+  @Override
+  public void close() throws IOException {
+    clients.close();
   }
 
-  private ThriftHiveMetastore.Client getClient() {
-    final URI metastoreUri = URI.create(MetastoreConf.getAsString(conf, THRIFT_URIS));
-    final int socketTimeOut = (int) MetastoreConf.getTimeVar(conf, CLIENT_SOCKET_TIMEOUT, TimeUnit.MILLISECONDS);
+  private ThriftHiveMetastore.Client newClient() {
```
Comment (rdblue) [inline hive/src/main/java/org/apache/iceberg/hive/HiveTables.java]: TODO: this is no longer used and can be removed.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 3]
```diff
@@ -126,7 +127,7 @@ public void commit(TableMetadata base, TableMetadata metadata) {
       // TODO add lock heart beating for cases where default lock timeout is too low.
       Table tbl;
       if (base != null) {
-        tbl = metaStoreClient.get_table(database, tableName);
+        tbl = metaClients.run(client -> client.getTable(database, tableName));
```
Comment (aokolnychyi) [inline hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: One question that is not related to this change. Am I correct that we don't refresh the metadata after we require the lock?

File: hive/src/main/java/org/apache/iceberg/hive/HiveTables.java [hunk 4]
```diff
@@ -70,12 +71,17 @@ public Table load(String tableIdentifier) {
 
   @Override
   public BaseMetastoreTableOperations newTableOps(Configuration conf, String database, String table) {
-    return new HiveTableOperations(conf, getClient(), database, table);
+    return new HiveTableOperations(conf, clients, database, table);
+  }
+
+  @Override
+  public void close() throws IOException {
```
Comment (aokolnychyi) [inline hive/src/main/java/org/apache/iceberg/hive/HiveTables.java]: +1 for this

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java [hunk 5]
```diff
@@ -126,93 +116,91 @@ public void cleanup() {
...
     return new TThreadPoolServer(args);
   }
 
-  private void setupDB(String dbURL) throws SQLException, IOException {
+  private static void setupDB(String dbURL) throws SQLException, IOException {
     Connection connection = DriverManager.getConnection(dbURL);
     ScriptRunner scriptRunner = new ScriptRunner(connection, true, true);
 
-    URL hiveSqlScript = getClass().getClassLoader().getResource("hive-schema-3.1.0.derby.sql");
+    URL hiveSqlScript = HiveTableBaseTest.class.getClassLoader().getResource("hive-schema-3.1.0.derby.sql");
```
Comment (aokolnychyi) [inline hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: Do we need to change the script file as well?

File: hive/src/test/java/org/apache/iceberg/hive/HiveTablesTest.java [hunk 6]
```diff
@@ -113,20 +112,22 @@ public void testConcurrentFastAppends() {
...
         for (int numCommittedFiles = 0; numCommittedFiles < 10; numCommittedFiles++) {
-          long commitStartTime = System.currentTimeMillis();
-          table.newFastAppend().appendFile(file).commit();
-          long commitEndTime = System.currentTimeMillis();
-          long commitDuration = commitEndTime - commitStartTime;
-          try {
-            TimeUnit.MILLISECONDS.sleep(200 - commitDuration);
-          } catch (InterruptedException e) {
-            throw new RuntimeException(e);
+          while (barrier.get() < numCommittedFiles * 2) {
```
Comment (aokolnychyi) [inline hive/src/test/java/org/apache/iceberg/hive/HiveTablesTest.java]: Looks better!

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  protected abstract C newClient();
+
+  private C get() throws InterruptedException {
+    Preconditions.checkArgument(!closed, "Cannot get a client from a closed pool");
+    while (true) {
+      if (!clients.isEmpty() || currentSize < poolSize) {
+        synchronized (this) {
+          if (!clients.isEmpty()) {
```
Comment (rdsr) [inline hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: It seems we are writing a lot of low level thread safe code ourselves. I'm just curious to know your thoughts whether a `LinkedBlockingQueue` would have made this easier?

File: build.gradle [hunk 8]
```diff
@@ -24,7 +24,7 @@ buildscript {
     maven { url  "http://palantir.bintray.com/releases" }
   }
   dependencies {
-    classpath 'com.github.jengelman.gradle.plugins:shadow:2.0.0'
```
[the remark this answers is not in the document]
Comment (rdblue) [reply build.gradle]: Yes, it was.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java (hunk 7 above)
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: I started out using one, but it doesn't work for the case where there is no available client and this should create a new one. In order to ensure that threads don't create a client in parallel, the check and the create need to happen inside a synchronized block. At that point, there's no benefit to using a concurrent queue.

**Comment by manishmalhotrawork** [comment]
@rdblue thanks for making these changes !

@rdblue @Parth-Brahmbhatt 
We are also looking to test Presto-sql + Iceberg Plugin. (https://github.com/prestosql/presto/pull/458)
Though Im wondering, as Presto-sql master is supporting Hive 3.x and not Hive 1.x.
So, the Iceberg version used in Presto's PR should be supporting Hive 3.x ?

Any thoughts on that direction as well? 

As @aokolnychyi and me wants to enable Iceberg usage from Spark and Presto.

Will it be like, for Spark we have to use Iceberg compatible with Hive 1.x version and for Presto use Iceberg compatible with Hive 3.x?

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Deque;
+import java.util.function.Consumer;
+
+abstract class ClientPool<C, E extends Exception> implements Closeable {
```
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: Not yet. I thought it may be useful in the future and shouldn't be tied to Hive, but I'm not sure there will be other uses so I kept it package-private for now.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  private volatile int currentSize;
+  private boolean closed;
+  private int runs = 0;
+
+  ClientPool(int poolSize, Consumer<C> reconnectFunc, Class<E> reconnectExc, Consumer<C> closeFunc) {
+    this.poolSize = poolSize;
+    this.reconnectFunc = reconnectFunc;
+    this.reconnectExc = reconnectExc;
+    this.closeFunc = closeFunc;
+    this.clients = Lists.newLinkedList();
```
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: Will do.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  interface Action<R, C, E extends Exception> {
+    R run(C client) throws E;
+  }
+
+  public <R> R run(Action<R, C, E> action) throws E, InterruptedException {
+    runs += 1;
+    C client = get();
+    try {
+      return action.run(client);
+
```
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: We don't need it, I just don't want the code to be too dense that it is hard to read.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    } catch (Exception exc) {
+      if (reconnectExc.isInstance(exc)) {
```
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: It depends on other uses. This is fine for now, so I'd prefer to keep it simple until we have more uses.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+    } finally {
+      release(client);
+    }
+  }
+
+  protected abstract C newClient();
+
+  private C get() throws InterruptedException {
+    Preconditions.checkArgument(!closed, "Cannot get a client from a closed pool");
```
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: You're right, I'll fix it.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 3 above)
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: This line is after the lock is acquired on line 125.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 1 above)
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: I was wrong. This needs to be Exception to include the TException thrown if unlock fails.

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 5 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: I made a small update so this one works with the 1.2.1 code.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 3 above)
Comment (aokolnychyi) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: Correct, but the refresh happens on line 111, right? Looks like we don't refresh after acquiring the lock, which means the table metadata can be changed after the refresh but before we acquire the lock?

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 14]
```diff
@@ -63,13 +60,13 @@
 public class HiveTableOperations extends BaseMetastoreTableOperations {
```
Comment (aokolnychyi) [inline hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: @rdblue do we need to keep the comment about thread-safety on line 57? The pool should resolve this issue.

**Comment by electrum** [comment]
@manishmalhotrawork The Presto Hive connector uses the Hive 3.x libraries, mainly to have access to the latest Parquet version, but that's an internal implementation detail. Today, Presto works with Hive 1.x and 2.x, and we have no plans to drop support for older versions. 

Presto does not yet support Hive 3.x due to backwards incompatible changes around transactional tables, bucketing, security, and timestamps. It seems that Hive 3.x is actually incompatible with every other tool in the ecosystem, including over Hive versions. Spark support seems to be implemented by running the queries in HiveServer2 (rather than native Spark).

One of the reasons why we (Presto developers and community) are so excited about Iceberg is that it frees users from the unwritten convention for Hive tables and provides a strong specification that all the tools in the ecosystem can implement. I'm also personally excited about the ability to replace the metastore with DynamoDB or similar, allowing users to run queries over data stored in cloud object stores without any additional dependencies.

**Comment by manishmalhotrawork** [comment]
@electrum thanks a lot !

trying to put together my understanding, thanks for the patience.

Presto master is using [Hive 3.x](https://github.com/prestosql/presto/blob/master/pom.xml#L426)
But it's not ready to use, for the reasons you mentioned above.
And I believe same thing will be true for Iceberg plugin for now.

So, to use Iceberg from Presto:
Use Presto-sql release - tag-300, was the last release using [Hive 1.x](https://github.com/prestosql/presto/blob/300/pom.xml#L410), so that should be used ?
and, once Iceberg is good with Hive 1.x, we should use that and use from Presto-Iceberg plugin.

**Comment by vrozov** [comment]
@electrum Can you please provide more details on the dependency between Hive and Parquet version in Presto? Does Presto use hive to read parquet files? My understanding was that Presto uses parquet libraries directly. Are Hive 3.x client libraries backward compatible with 1.x and/or 2.x Hive metastore?

**Comment by electrum** [comment]
@manishmalhotrawork No, you should use the latest release, currently `308`, and the Iceberg plugin should be based on `master`. Presto supports Hive 1.x and 2.x but not 3.x. Ignore the version of the Hive libraries it uses -- that is an internal implementation detail and has no bearing on what version of Hive it supports.

@vrozov Presto uses the Hive and Parquet libraries to read and write Parquet. For reading, most of the reading happens in Presto's optimized Parquet reader, which uses the lower level parts of the Parquet libraries. For writing, it uses `MapredParquetOutputFormat` (directly as a library). This dependency is why we need to upgrade the Hive and Parquet libraries together (older versions of the Hive library code would break with the newer Parquet library versions).

The Hive 3.x library code is backwards compatible with older Hive metastore versions because it is a Thrift client (and they did not make backwards compatible changes to the Thrift IDL / wire protocol).

**Supported**: Hive 1.x and 2.x
**Not Supported**: Hive 3.x

**Comment by electrum** [comment]
Another point, in case I wasn't clear above, Presto has extensive integration tests that run against HDP 2.6 (Hive 2.x) in Docker and run as part of every Travis build.

If you want to have a more realtime discussion, please feel free to join our Slack. I'm `david` on there. Invite link here: https://prestosql.io/community.html

**Comment by vrozov** [comment]
@electrum @rdblue If Hive 3.x libraries are backward compatible with Hive 1.x and Hive 2.x metastore, why is it necessary to downgrade Hive version in this PR? Should it be sufficient to shade and relocate Hive dependency inside iceberg runtime?

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 5 above)
Comment (rdsr) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: Would it then make sense to drop the `3.1.0` version suffix?

**Comment by rdblue** [comment]
@vrozov, the reason to downgrade here is to avoid accidentally using features that weren't in older versions. For example, the API in 3.0 includes a catalog name that didn't exist before. Calling those methods won't work in 1.2.1, which will be the runtime version for Spark. But using 1.2.1 methods works because 3.0 is compatible.

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 5 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: I'm reluctant to because it would make the history of the file harder to follow. A rename would obscure the change because it is a whole-file delete and a whole-file add.

I'd say that the easiest thing is to keep the file name as it is. That way we know the version it was based on.

File: build.gradle [hunk 15]
```diff
@@ -78,7 +78,7 @@ subprojects {
     avroVersion = '1.8.2'
     orcVersion = '1.4.2'
     parquetVersion = '1.10.0'
-    hiveVersion = '3.1.0'
+    hiveVersion = '1.2.1'
```
Comment (vrozov) [inline build.gradle]: Please see my comment regarding shading, should it be sufficient to shade and relocate hive libraries to avoid conflict with Spark?

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 3 above)
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: Line 111 (`base != current()`) only checks the currently-known metadata. If before a refresh the base metadata is already out of date, it fails early.

After the lock is acquired, table metadata is refreshed using this call to `getTable` on line 130 (in the latest version). On line 148, this checks the metadata location that was fetched to ensure that it is the one that was expected.

Comment (aokolnychyi) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: You are right, we handle it correctly.

File: build.gradle (hunk 15 above)
Comment (vrozov) [reply build.gradle]: @rdblue OK, make sense. Note that for a project that uses both presto and iceberg it is quite confusing that presto upgrades to 3.x hive and iceberg downgrades to 1.x.

File: build.gradle [hunk 16]
```diff
@@ -24,7 +24,7 @@ buildscript {
     maven { url  "http://palantir.bintray.com/releases" }
   }
   dependencies {
-    classpath 'com.github.jengelman.gradle.plugins:shadow:5.0.0'
+    classpath 'com.github.jengelman.gradle.plugins:shadow:2.0.0'
```
[the remark this answers is not in the document]
Comment (rdblue) [reply build.gradle]: It isn't. This PR and another one updated it to 5.0 at the same time. I'm trying to remove the duplication.

[the remark this answers is not in the document]
Comment (vrozov) [reply build.gradle]: Do you refer to PR #151? This change will introduce the issue fixed in #151 back.

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 5 above)
Comment (rdsr) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: Makes sense

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java [hunk 17]
```diff
@@ -126,93 +116,91 @@ public void cleanup() {
...
-    IHMSHandler handler = RetryingHMSHandler.getProxy(hiveConf, baseHandler, true);
-    final TTransportFactory transportFactory = new TTransportFactory();
-    final TSetIpAddressProcessor<IHMSHandler> processor = new TSetIpAddressProcessor<>(handler);
+  private static TServer thriftServer() throws Exception {
+    TServerSocket socket = new TServerSocket(0);
+    HiveTableBaseTest.hiveConf = hiveConf(new Configuration(), socket.getServerSocket().getLocalPort());
+    HiveConf serverConf = new HiveConf(hiveConf);
+    serverConf.set(HiveConf.ConfVars.METASTORECONNECTURLKEY.varname, "jdbc:derby:" + getDerbyPath() + ";create=true");
+    HiveMetaStore.HMSHandler baseHandler = new HiveMetaStore.HMSHandler("new db based metaserver", serverConf);
+    //IHMSHandler handler = RetryingHMSHandler.getProxy(serverConf, baseHandler, false);
```
Comment (rdsr) [inline hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: remove?

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java [hunk 18]
```diff
@@ -126,93 +116,91 @@ public void cleanup() {
...
   }
 
-  private void setupDB(String dbURL) throws SQLException, IOException {
+  private static void setupDB(String dbURL) throws SQLException, IOException {
     Connection connection = DriverManager.getConnection(dbURL);
     ScriptRunner scriptRunner = new ScriptRunner(connection, true, true);
 
-    URL hiveSqlScript = getClass().getClassLoader().getResource("hive-schema-3.1.0.derby.sql");
+    URL hiveSqlScript = HiveTableBaseTest.class.getClassLoader().getResource("hive-schema-3.1.0.derby.sql");
     Reader reader = new BufferedReader(new FileReader(new File(hiveSqlScript.getFile())));
```
Comment (rdsr) [inline hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: nit: doesn't seem like the reader is closed. But I guess it doesn't make too much of a diff. in a test

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: I don't think I agree about using `AutoCloseable`. Two reasons:
1. Close will close client connections, which do release IO streams. So this is an IO stream close.
2. Even if `AutoCloseable` were a better fit for exceptions, `Closeable` is the one more often used. I even wrote the initial version of close by checking whether a client implemented `Closeable` to call close. That wouldn't work with `AutoCloseable`. I don't think people actually think to use `AutoCloseable` in those cases, so implementing `Closeable` is more likely to work.

File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 14 above)
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: You're right. I've removed it.

File: build.gradle (hunk 16 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply build.gradle]: This is an artifact of rebasing. I'll not commit this with a change that reverts #151.

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 18 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: Doesn't matter much, but I'll fix it anyway.

File: hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java (hunk 17 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java]: This was an unnecessary change I introduced while trying to find out why the tests were failing. I'll revert it.

File: build.gradle (hunk 16 above)
[the remark this answers is not in the document]
Comment (vrozov) [reply build.gradle]: It was not clear from the PR that the change won't be committed.

[the remark this answers is not in the document]
Comment (rdblue) [reply build.gradle]: It has been removed.

File: hive/src/main/java/org/apache/iceberg/hive/ClientPool.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vrozov) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: 1. The difference between `Closeable` and `AutoCloseable` is only in the type of exception thrown by the `close()` method. In the case of `ClientPool`, `close()` does not throw `IOException`, so it is closer to `AutoCloseable`.
1. `ClientPool` does not close IO streams by itself. It releases resources (connections).
1. It is arguable what is more often used as every class that implements `Closeable` also implements `AutoCloseable`.
1. The question is what interface `ClientPool` implements. An individual resource/client/connection that `ClientPool` manages may or may not implement `Closeable` or `AutoCloseable` at all. For example in case of `HiveMetaStoreClient` it implements `IMetaStoreClient` that is neither of those.

[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: You're right that this doesn't currently wrap a `Closeable`, but I think that because `Closeable` is the generally expected interface, it is better to implement that one. `AutoCloseable` didn't exist until Java 7 and `Closeable` extends `AutoCloseable`, so the one that typically gets used is `Closeable`. It may well be more correct to use `AutoCloseable`, but it's a minor distinction and this works in more cases.

[the remark this answers is not in the document]
Comment (vrozov) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: I agree that it's a minor distinction (so was the suggestion to consider switching to `AutoCloseable`, not a request for the code change), but I don't agree that `Closeable` works in *more* cases.

[the remark this answers is not in the document]
Comment (rdblue) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: Thanks for the context. By more cases, I mean this: all instances of `Closeable` are `AutoCloseable` and people are used to handling `Closeable` and may not check for `AutoCloseable`. So there are cases where callers expect `Closeable` and returning an `AutoCloseable` won't work. I think the logic holds because all `Closeable` instances are `AutoCloseable`, but not vice versa. Anyway, it's minor and this isn't even public so it doesn't matter.

[the remark this answers is not in the document]
Comment (vrozov) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: I don't agree with that logic. It is incorrect to expect that everything is `Closeable`. According to java doc, *A `Closeable` is a source or destination of data that can be closed.* `ClientPool` is neither a source or a destination of data. It is a managed collection of clients. IMO, it is incorrect for an object to pretend that it is a `cat` when it is an `animal` due to other objects being only familiar how to deal with cats :).

Yes, it's a minor comment, not a blocker for the PR to be merged.

[the remark this answers is not in the document]
Comment (electrum) [reply hive/src/main/java/org/apache/iceberg/hive/ClientPool.java]: An important distinction is that `Closeable` is required to be idempotent but `AutoCloseable` is not.

**Comment by rdblue** [comment]
Thanks for the reviews and comments, everyone! I've merged this. Next step is to add support for Hive tables to the Spark integration.
